rosbag2/src/rosbag2/writer.cpp

```diff
@@ -90,7 +111,52 @@ void Writer::write(std::shared_ptr<SerializedBagMessage> message)
     throw std::runtime_error("Bag is not open. Call open() before writing.");
   }
 
+  // Update the message count for the Topic.
+  ++topics_[message->topic_name].message_count;
+  ++message_count_;
+
+  start_time_ = std::min(start_time_, message->time_stamp);
+  end_time_ = std::max(end_time_, message->time_stamp);
+
   storage_->write(converter_ ? converter_->convert(message) : message);
 }
 
+bool Writer::should_split_bagfile() const
+{
+  if (max_bagfile_size_ == rosbag2_storage::storage_interfaces::MAX_BAGFILE_SIZE_NO_SPLIT) {
+    return false;
+  } else {
+    return storage_->get_bagfile_size() > max_bagfile_size_;
+  }
+}
+
+rosbag2_storage::BagMetadata Writer::generate_metadata_() const
+{
+  auto time_of_first_message = std::chrono::nanoseconds(start_time_);
+  auto time_of_last_message = std::chrono::nanoseconds(end_time_);
+
+  rosbag2_storage::BagMetadata metadata{};
+
+  // Only populate metadata if storage exists
+  if (storage_) {
+    metadata.storage_identifier = storage_->get_identifier();
+
+    metadata.relative_file_paths = relative_file_paths_;
+    metadata.duration = time_of_last_message - time_of_first_message;
+    metadata.starting_time =
+      std::chrono::time_point<std::chrono::high_resolution_clock>(time_of_first_message);
+    metadata.message_count = message_count_;
+
+    metadata.bag_size = rosbag2_storage::FilesystemHelper::calculate_directory_size(uri_);
```

Comment:
I understand that we used that function before, but I actually think now that we can't solely rely on the directory size. There could (even though shouldn't) be files in the same directory which are essentially not related to the payload of the bag file. 
I think it's preferable to iterate over the individual relative paths and ask for the filesize: https://github.com/ros2/rosbag2/blob/master/rosbag2_storage/include/rosbag2_storage/filesystem_helper.hpp#L132-L137

Reply:
Makes sense, shouldn't be too hard since `Writer` already knows of all of the bagfiles.